_drafts/conviva-dag-process-time-prod-issue-draft.md

```diff
@@ -0,0 +1,69 @@
+
+In the context of this, the earlier graph about the context switches spiking during the incident time makes some sense. The `ReadGuard` in the hot path of the flamegraph was responsible for handling reads from various threads and each thread would increment and decrement the counter.
+
+Now, one important thing about the hashmap in the type registry is that it is *almost* read-only. That is, it is initialized with some types on start-up and then only updated when a new type is seen but that rarely ever happened. However, on a critical path, it would keep checking to see if the type was already registered which is where the atomic increments and decrements were occurring.
+
+Now, the question was what to do about this. In the [flashmap documentation](https://crates.io/crates/flashmap), the performance comparison shows that in the read-heavy scenario [`dashmap`](https://github.com/xacrimon/dashmap) performed better in terms of both latency & throughput. Unfortunately, replacing `flashmap` with `dashmap` did nothing to fix the performance problems. In fact, the flamegraphs turned out to be worse in the same situation with `dashmap`.
+
+![](/assets/img/conviva/rtve/perf_incident_traffic_dashmapv2.svg)
+
+Finally, [Evan Chan](https://www.linkedin.com/in/evanfchan/), a Principal Engineer at Conviva, implemented an [ArcSwap](https://github.com/vorner/arc-swap) based solution and the flamegraph for that was significantly better and the CPU load dropped to 40% in the perf environment.
+
+## Post Mortem
+
+So, ArcSwap finally fixed it but it is worthwhile to understand why exactly that worked where the concurrent hash maps had failed. To understand this better, look at this section below from a great PDF titled [What every systems programmer should know about concurrency](https://assets.bitbashing.io/papers/concurrency-primer.pdf) by [Matt Kline](https://github.com/mrkline).
+
+![](/assets/img/conviva/rtve/cache_line_ping_pong.png)
+
+Every CPU has a cache, and every cache requires loading data in from RAM. However, when you have a shared variable like a read counter that is being used by all the cores there is a critical problem under high contention. Each core is going to attempt to increment/decrement the same read counter (its a shared counter), and each modification causes cache invalidation because [cache coherence](https://en.wikipedia.org/wiki/Cache_coherence) dictates that any change to a cache line requires re-fetching from main memory. This also ties in with the context switching graph we saw earlier, which showed a spike in context switches during the incident.
```

Comment:
I do agree the cache invalidation is potentially behind why atomic counters are slow.  However, I think if it was me I would focus on how and why concurrent hash maps need atomic counters, and how that invariably leads to contention.  

ArcSwap is not a concurrent hash map.  It is a very specialized AtomicRef, designed for occasional swaps, but the entire ref is updated.  It happens to work for our situation of a type registry.

It seems this begs for a blog post about LeanBox, which we plan to open source.  I guess we should open source it and blog about it, and that's probably a separate blog post.

Reply:
@velvia Yeah, did this. Expanded on the differences between something like `ArcSwap` and a concurrent hash map and why it makes sense for our use case. 

And yes, definitely need a blog post on leanbox to provide more context. Also, I'm fairly sure that would get a lot of attention on places like HackerNews and Lobsters